### PR TITLE
Fix coding style for mock

### PR DESCRIPTION
**Pull request by hlts2**
Signed-off-by: hlts2 <hiroto.funakoshi.hiroto@gmail.com>

### Description:

I fixed coding style for mock and GitHub action.

WHY
- In Vald, There are a lot of dependencies between libraries.
As a result, the complexity of the test has increased, and it has become more difficult to determine whether or not to mock dependent objects and how to create mock.

- if there are a lot of review dog comment, we can not see pr page.  (review dog output warning because document has alot of example code. but we want to ignore it)

WHAT 
- Add  document for mock

- Add `TYPOGRAPY` to [action yaml](https://github.com/vdaas/vald/blob/master/.github/workflows/reviewdog-markdown.yml)

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [x] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: docs/contributing/coding-style.md [hunk 1]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+            }
+            ......
+        }
+    }
+
+    ......
+
+    ```
+
+#### Mock rick
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1059]: I think this section should before the example
```suggestion
#### Risk
```

File: docs/contributing/coding-style.md [hunk 2]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+
+- Incomplete implementation
+- IO
+  - Network access, disk operation, etc.
+- Hardware dependent
+  - CPU, memory usage, disk IO, etc.
+- Difficult to create error of dependent object (when we will write error test case)
+- Difficult to initialize
+  - Random number and time, file IO initialization, environment dependent, etc.
+- Test result may change
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:986]:
```suggestion
- Test result may change in runtime
```

File: docs/contributing/coding-style.md [hunk 3]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+  - Cases where the implementation and test code are not changed but the test result changes
+  - System call dependent
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:988]:
```suggestion
  - System call inside implementation
```

File: docs/contributing/coding-style.md [hunk 4]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
         type want struct {
             // generated test code
     ```
+### Mock
+
+In Vald, we use a lot of external library, there are a lot of dependencies between libraries.
+
+As a result, the complexity of the test has increased, and it has become more difficult to determine whether or not to mock dependent objects.
+
+#### Cases where you can use mock
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:974]:
```suggestion
#### Condition
```

File: docs/contributing/coding-style.md [hunk 5]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- Hardware dependent
+  - CPU, memory usage, disk IO, etc.
+- Difficult to create error of dependent object (when we will write error test case)
+- Difficult to initialize
+  - Random number and time, file IO initialization, environment dependent, etc.
+- Test result may change
+  - Cases where the implementation and test code are not changed but the test result changes
+  - System call dependent
+
+#### How to create mock
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:990]:
```suggestion
#### Implementation
```

File: docs/contributing/coding-style.md [hunk 6]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+        return e.encoder.Encode(obj)
+    }
+    ```
+
+    And there are 4 rules of mock code.
+    
+    - File location is same pacakge as mock target.
+    - File name is `〇〇_mock.go`
+    - Structure name are `Mock{Interface name}`
+    - Method injected from test code is `{Method name}Func`
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1020]: We should move these rules to another section

File: docs/contributing/coding-style.md [hunk 7]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+    func (e *encoder) Encode(obj interface{}) ([]byte, error) {
+        return e.encoder.Encode(obj)
+    }
+    ```
+
+    And there are 4 rules of mock code.
+    
+    - File location is same pacakge as mock target.
+    - File name is `〇〇_mock.go`
+    - Structure name are `Mock{Interface name}`
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1019]:
```suggestion
    - Struct name is `Mock{Interface name}`
```

File: docs/contributing/coding-style.md [hunk 8]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+    }
+
+    func (e *encoder) Encode(obj interface{}) ([]byte, error) {
+        return e.encoder.Encode(obj)
+    }
+    ```
+
+    And there are 4 rules of mock code.
+    
+    - File location is same pacakge as mock target.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1017]:
```suggestion
    - Same package as the mock target.
```

File: docs/contributing/coding-style.md (hunk 6 above)
Comment (kevindiu) [inline docs/contributing/coding-style.md:1020]:
```suggestion
```

File: docs/contributing/coding-style.md [hunk 9]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+    func (m *MockEncoder) Encode(obj interface{}) ([]byte, error) {
+        return m.EncodeFunc(obj)
+    }
+
+    ```
+
+    The following is an example implementation of test code to create the mock object and mock the implementation.
+
+    ```go
+
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1040]:
```suggestion
```

File: docs/contributing/coding-style.md [hunk 10]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+        }
+    }
+
+    ......
+
+    ```
+
+#### Mock rick
+
+- We don't know if it's really correct.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1061]: can you explain more? I cannot understand what it means

**Review by kevindiu — CHANGES_REQUESTED** [review]

File: docs/contributing/coding-style.md (hunk 6 above)
Comment (hlts2) [reply docs/contributing/coding-style.md:1020]: I think we should write about mock method format.
what do you think?

Comment (kevindiu) [reply docs/contributing/coding-style.md]: I think so.
But it is not the method name, it is the function implementation of the functions inside the mock object

File: docs/contributing/coding-style.md [hunk 11]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+#### Implementation
+
+1. Basic mock example.
+
+    For example, we decided to mock the following implementation `Encoder`.
+
+    ```go
+    package json
+
+    type Encoder struct {
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:999]: is it an interface?

File: docs/contributing/coding-style.md [hunk 12]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+
+    ```
+
+#### Rules of mock code
+
+There are 4 rules of mock code.
+
+- Same package as the mock target.
+- File name is `〇〇_mock.go`
+- Struct name is `Mock{Interface name}`
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1057]: I think we should describe the rule first before the example

File: docs/contributing/coding-style.md [hunk 13]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+        }
+    }
+
+    ......
+
+    ```
+
+#### Rules of mock code
+
+There are 4 rules of mock code.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1053]:
```suggestion
There are 3 rules of mock code.
```

Comment (kevindiu) [inline docs/contributing/coding-style.md:1053]: Please move this category inside `implementation` section
```suggestion
The implementation of mock object should be:
```

File: docs/contributing/coding-style.md (hunk 12 above)
Comment (hlts2) [reply docs/contributing/coding-style.md:1057]: I think so too. thankyou

File: docs/contributing/coding-style.md (hunk 13 above)
Comment (hlts2) [reply docs/contributing/coding-style.md:1053]: thank you. 

by the way, It is different from the following opinion, so I would like you to unify it from now on.
https://github.com/vdaas/vald/pull/806#discussion_r513967570

File: docs/contributing/coding-style.md (hunk 11 above)
Comment (hlts2) [reply docs/contributing/coding-style.md:999]: thank you. I made a misstake. so I will fix it soon.

File: docs/contributing/coding-style.md [hunk 14]
```diff
@@ -965,3 +965,96 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+    }
+
+    ......
+
+    ```
+
+#### Risk
+
+- We don't know if the dependent object is correct.
+- We don't notice the change in dependency.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:1060]: I think it is better to move this section before ``implementation` section.
```suggestion
#### Risk

Before applying mock to the object, you should be aware of the following risks.

- We do not know if the dependent object is correctly implemented or not.
- We cannot notice the change in dependency.
```

File: docs/contributing/coding-style.md [hunk 15]
```diff
@@ -965,3 +965,96 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
     ```
+### Mock
+
+In Vald, we use a lot of external library, there are a lot of dependencies between libraries.
+
+As a result, the complexity of the test has increased, and it has become more difficult to determine whether or not to mock dependent objects.
+
+#### Condition
+
+When a dependent object has the following feature, you can decide to mock the dependent.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:976]:
```suggestion
When a dependent object has the following factor, you can decide to mock the dependent object.
```

File: docs/contributing/coding-style.md (hunk 14 above)
Comment (hlts2) [reply docs/contributing/coding-style.md]: thank you.
It looks good to me.
I will fix it soon

**Review by kevindiu — DISMISSED** [review]

File: docs/contributing/coding-style.md [hunk 16]
```diff
@@ -965,3 +965,98 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+                    },
+                },
+            }
+            ......
+        }
+    }
+
+    ......
+
+    ```
```
Comment (vankichi) [inline docs/contributing/coding-style.md:1062]:
```suggestion
  For example, we decided to mock the following implementation `Encoder`.

  ```go
  package json

  type Encoder interface {
      Encode(interface{}) ([]byte, error) 
  }

  ```

  ```go
  type encoder struct {
      encoder json.Encoder
  }

  func (e *encoder) Encode(obj interface{}) ([]byte, error) {
      return e.encoder.Encode(obj)
  }
  ```

  The following is an example of mock implementation:

  ```go
  package json

  type MockEncoder struct {
      EncoderFunc func(interface{}) ([]byte, error)
  }

  func (m *MockEncoder) Encode(obj interface{}) ([]byte, error) {
      return m.EncodeFunc(obj)
  }

  ```

  The following is an example implementation of test code to create the mock object and mock the implementation.

  ```go
  tests := []test {
      {
          name: "returns (byte{}, nil) when encode success"
          fields: fields {
              encoding: &json.MockEncoder {
                  EncoderFunc: func(interface{}) ([]byte, error) {
                      return []byte{}, nil
                  },
              },
          },
          ......
      }
  }

  ```

```

**Review by vankichi — COMMENTED** [review]
@hlts2 commented

File: docs/contributing/coding-style.md [hunk 17]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+
+#### Condition
+
+When dependencies have the following factor, you can decide to mock the dependencies.
+
+- Incomplete implementation
+- I/O
+  - e. g. Network access, disk operation, or etc.
+- Hardware dependent
+  - e. g. CPU, Memory usage, disk I/O, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:982]:
```suggestion
  - e. g. CPU, Memory usage, disk I/O, etc.
```

File: docs/contributing/coding-style.md [hunk 18]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+
+As a result, due to the more complexity of the test, it has become more difficult to determine whether or not to mock dependencies.
+
+#### Condition
+
+When dependencies have the following factor, you can decide to mock the dependencies.
+
+- Incomplete implementation
+- I/O
+  - e. g. Network access, disk operation, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:980]:
```suggestion
  - e. g. Network access, disk operation, etc.
```

File: docs/contributing/coding-style.md [hunk 19]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- Hardware dependent
+  - e. g. CPU, Memory usage, disk I/O, or etc.
+- Difficult to create error of dependencies
+- Difficult to initialize
+  - e. g. Random number and time, file IO initialization, environment dependent, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:985]:
```suggestion
  - e. g. Random number and time, file IO initialization, environment dependent, etc.
```

File: docs/contributing/coding-style.md [hunk 20]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- Test result may change in each runtime
+  - e. g. Only test result may change in each runtime, System call inside implementation, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:987]:
```suggestion
  - e. g. Only test result may change in each runtime, System call inside implementation, etc.
```

File: docs/contributing/coding-style.md [hunk 21]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+
+#### Risk
+
+Before applying mock to the object, you should be aware of the following risks.
+
+- We **do not** know whether the dependencies is correctly implemented or not.
+- We cannot notice the change in dependencies.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:994]:
```suggestion
- We cannot notice the changes in dependencies.
```

File: docs/contributing/coding-style.md [hunk 22]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- Difficult to initialize
+  - e. g. Random number and time, file IO initialization, environment dependent, or etc.
+- Test result may change in each runtime
+  - e. g. Only test result may change in each runtime, System call inside implementation, or etc.
+
+#### Risk
+
+Before applying mock to the object, you should be aware of the following risks.
+
+- We **do not** know whether the dependencies is correctly implemented or not.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:993]:
```suggestion
- We **do not** know whether the dependencies are correctly implemented or not.
```

File: docs/contributing/coding-style.md [hunk 23]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- We cannot notice the change in dependencies.
+
+#### Implementation
+
+The implementation of mock object should be:
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:998]:
```suggestion
The implementation of the mock object should be:
```

File: docs/contributing/coding-style.md [hunk 24]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+
+As a result, due to the more complexity of the test, it has become more difficult to determine whether or not to mock dependencies.
+
+#### Condition
+
+When dependencies have the following factor, you can decide to mock the dependencies.
+
+- Incomplete implementation
+- I/O
+  - e.g. Network access, disk operation, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:980]:
```suggestion
  - e.g. Network access, disk operation, etc.
```

File: docs/contributing/coding-style.md [hunk 25]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- Hardware dependent
+  - e.g. CPU, Memory usage, disk I/O, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:982]:
```suggestion
  - e.g. CPU, Memory usage, disk I/O, etc.
```

File: docs/contributing/coding-style.md [hunk 26]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
...
+- Difficult to create error of dependencies
+- Difficult to initialize
+  - e.g. Random number and time, file I/O initialization, environment dependent, or etc.
```
Comment (kevindiu) [inline docs/contributing/coding-style.md:985]:
```suggestion
  - e.g. Random number and time, file I/O initialization, environment dependent, etc.
```

**Review by kevindiu — DISMISSED** [review]
LGTM

File: docs/contributing/coding-style.md [hunk 27]
```diff
@@ -965,3 +965,90 @@ We do not suggest to modify the generated code other than the `tests` variable,
         type want struct {
             // generated test code
     ```
+### Using Mock
+
+In Vald, we use a lot of external library, there are a lot of dependencies between libraries.
```
Comment (vankichi) [inline docs/contributing/coding-style.md:970]:
```suggestion
In Vald, we use a lot of external libraries, there are a lot of dependencies between libraries.
```

**Review by kevindiu — APPROVED** [review]

**Review by vankichi — APPROVED** [review]

**Pull request merged**